### PR TITLE
feat(commands): create Go init hook

### LINKED ISSUE #1209: flox init reads go.mod and suggests installation/config
# Acceptance Criteria

**Given** the project has `go.mod` file
**And** i'm in the project directory
**When** I `flox init`
**Then** I see an interactive prompt

```
Flox detected a go.mod

Go projects typically need:
* Go
* A shell hook to apply environment variables

Would you like Flox to apply the standard Go environment? 
You can always revisit the environment's declaration with 'flox edit'

>Yes 
No
Show environment manifest

```

# Details

Suggested installs

```
go
```

Suggested shell hook
```
[profile]
common = """
export GOENV="$FLOX_ENV_CACHE/goenv"
"""
```

### PR DESCRIPTION
**Pull request by oxcabe**
## Proposed Changes

This PR implements an init hook for the Go programming language.

Closes #1209.

## Release Notes

Running `flox init` in a directory containing `go.mod` or `go.work` files now prompts to apply a standard Go environment.

### CODE REVIEW

**Comment by oxcabe** [comment]
Thanks for your review! I have been mixing `python` and `node` hook approaches, so it's good to know that the former is preferred.

I will address your change suggestions ASAP, they are highly appreciated as well.

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:53]: > question: is it possible for GoWorkspace::try_new_from_path to Err but GoModule::try_new_from_path to succeed?

Yes, this could technically happen. Since the only source of Err is `fs::read_to_string`, possible errors are those defined in [`std::io::ErrorKind`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html). So, if one of these occurs, either the I/O is failing or the contents in `go.work` are corrupted.

From the [Go Modules Reference](https://go.dev/ref/mod#workspaces):

> A command determines whether it is in a workspace context by first examining the GOWORK environment variable. [...]. If it is empty or not provided, the command will search the current working directory, and then successive parent directories, for a file go.work. If a file is found, the command will operate in the workspace it defines; otherwise, the workspace will include only the module containing the working directory.

My personal interpretation (in relation to our implementation) is that, if there is a `go.work` file and the method returns Err, there is something strange going on with the project.

If the `go.work` file content syntax is non-compliant to the spec, the Go compiler complains, regardless of the existence of any `go.mod` files:

```
go: reading go.work: /home/oxcabe/misc/cilium-ebpf/go.work:1: unknown directive: helloworld
```

The only case where `go.mod` would apply is the one where `go.work` is not a file but a directory (and that should be checked).

> If so, bubbling up the error for the former should probably be deferred until after GoModule was checked.

The only issue I see with this is that, if `go.work` is unintentionally non-compliant (I'd even say corrupted), we might be packaging wrong dependencies for the project by checking the `go.mod` when we shouldn't. Although, after all, flox is not a linter nor a static checker for project structure and, in my opinion, should not be held responsible for any unexpected behavior caused by corrupted projects.

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:53]: Some extra considerations for the hook implementation:

* Should we check the `GOWORK` environment variable, since it might point to a workspace in a different directory? 

and

* Does flox require that users run `flox init` in the top level directory of projects? If it doesn't the hook might also need to look for `go.work` files in parent directories.

Comment (ysndr) [reply cli/flox/src/commands/init/go.rs:53]: I guess if go tooling ignores `go.mod` in the presence of `go.work` or `GOWORK` we can likely do the same.



> Does flox require that users run flox init in the top level directory of projects?

So far we didn't intend to require envs to be in the project root / repository root (kind of a nogo in monorepos...).

The other integrations currently do not look upwards but doing so is not implausible.
flox itself looks up to the repo root for an environment, I think it may be good to limit lookups of `go.work` similarly.

For GOWORK i could go either way: limit that to the repo/context of the environment (ie.e. GOWORK would be respected if GIT_ROOT is a common prefix of GOWORK and PWD) respect it generally or reject it.

especially GOWORK and go.work to some extend seem to get rather complicated to generate the right hooks for, that are actually portable.
I'd start with a more limited solution first and look at improving that part separately if that works.

with that, 

> Should we check the GOWORK environment variable, since it might point to a workspace in a different directory?

I think we should detect that it's there and only install go packages, leaving a note or warning that we dont provide portable hooks for that situation yet, no need to get too fancy.

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:102]: That comment was added before you introduced me to `format_customization`, which was the functionality I was looking for at that time. As you mention, it has nothing to do here and should have been removed already. Thanks for pointing that out!

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:234]: Exactly! I am skipping DRY here in order to start discussing where to relocate the shared functionality while still continuing to develop the Go hook.

I was worried about Go release versions [not being semver](https://github.com/golang/go/issues/32450) which could've implied extra complexity or incompatibilities, but then I noticed that `ProvidedVersion` in the Python hook is not tied to any versioning specs.

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [reply cli/flox/src/commands/init/go.rs]: I see no worries I was just wondering what you may have seen there.

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs]: I was just checking everything inside `cli/flox-rust-sdk/src/models/environment/` in order to understand environments, especially for files referencing manifests. `generations.rs` just happened to be the first file I noticed that seemed to have relevance, so I left that message as a "checkpoint".

I write most of my `TODO`, `XXX`... comments as reminders or references to things I've seen that I might need to revisit, and they should not make it to the merge commit in case the PR is accepted.

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [inline cli/flox/src/commands/init/go.rs:71]: Calling `Self::prompt_user` and `Self::get_init_customization` if `should_run` returns false would cause an invalid state, which should be unrepresentable. I propose refactoring this on a different PR to make the static checker aware of this.

File: cli/tests/environment-init.bats
Comment (mkenigs) [inline cli/tests/environment-init.bats:220]: nonblocking (no need to address this before we merge): we only test each activate method for one language, and then all the other languages we test in `tests/$language.bats` generally against just one activation method. We don't run the tests in `tests/` against every commit pushed to a PR since they tend to be on the slow side, whereas the ones in `cli/tests` we do run on every commit

Comment (oxcabe) [reply cli/tests/environment-init.bats:220]: Got it! I had already started with the `tests/go.bats` approach, but didn't push it since I wanted to see if this last commit made sense at all.

**Comment by ghudgins** [comment]
blind tried on `kubernetes/kubernetes` to see what it would do with the latest from this branch.

not sure if this is useful information or not! 😅 

image

**Comment by oxcabe** [comment]
> not sure if this is useful information or not! 😅

Definitely was useful! This also shows the need for adding proper integration test cases before it gets merged.

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [reply cli/flox/src/commands/init/go.rs:78]: Yes, this is a known limitation, which when ignored leads to unreachable/panic blocks which is not ideal, but ok, since its currently rather internal.
I wouldn't stress refactoring that here but would welcome a follow up discussion.

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:78]: I will open a PR for addressing this in the future.

**Comment by mkenigs** [comment]
It looks like that last commit isn't actually a merge commit - could you drop or revert that and then do a `git merge upstream/main`?

**Comment by oxcabe** [comment]
@mkenigs I have git misconfigured in some way, ended up resolving conflicts from the GitHub interface :sweat_smile:

**Comment by oxcabe** [comment]
Also - I am aware that some of the unit tests are broken as of the last commit.

**Comment by mkenigs** [comment]
> @mkenigs I have git misconfigured in some way, ended up resolving conflicts from the GitHub interface 😅

Looks good now! Took the diff from 3800 down to 700 lines haha

**Comment by oxcabe** [comment]
@mkenigs the tests I'm not getting to work depend on a "locked" flox environment:

```rust
let (flox, _) = flox_instance_with_global_lock();
```

The Go binary is retrieved correctly when I run the `flox init` command in a Go module based project, but the unit test fails with the same version.

Could it be that, while trying to retrieve the `relpath` for `go` packages, the globally locked instance can't seem to find and provide it?

File: cli/flox/src/commands/init/go.rs
Comment (mkenigs) [inline cli/flox/src/commands/init/go.rs:448]: With `_`, the temp dir will get dropped, so then subsequent operations will fail
```suggestion
        let (flox, _temp_dir_handle) = flox_instance_with_global_lock();
```

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:448]: I see, will change it for all instances :+1

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:448]: This solves my previous comment too :slightly_smiling_face:

**Comment by oxcabe** [comment]
Everything should be covered now (aside from documentation)

File: cli/flox/src/commands/init/go.rs
Comment (mkenigs) [inline cli/flox/src/commands/init/go.rs:289]: nonblocking: I think it's confusing having these implemented on `ProvidedVersion` when `ProvidedVersion` is used elsewhere for non-Go things

File: cli/flox/src/commands/init/go.rs
Comment (mkenigs) [inline cli/flox/src/commands/init/go.rs:438]: nonblocking: might make more sense to split out `get_version_string_from_content` tests from `from_module_system_content` tests

File: cli/flox/src/commands/init/go.rs
Comment (mkenigs) [inline cli/flox/src/commands/init/go.rs:324]: blocking: if I'm understanding https://go.dev/doc/modules/gomod-ref correctly, version 2 would satisfy a `go 1` constraint. Whereas `VersionReq` turns that into `^1`, which version 2 doesn't satisfy. No need to handle that, but can you at least add a comment saying we diverge slightly?

File: cli/flox/src/commands/init/go.rs
Comment (mkenigs) [inline cli/flox/src/commands/init/go.rs:311]: blocking:
Looks like the version specifier is mandatory. From https://go.dev/doc/modules/gomod-ref
```
Before Go 1.21, the directive was advisory only; now it is a mandatory requirement: Go toolchains refuse to use modules declaring newer Go versions.
```

So I think if we can't provide a Go version that satisfies the requirement, we shouldn't run the hook at all, because the go we provide wouldn't be able to compile the project. Ping @ghudgins

File: tests/go.bats
Comment (mkenigs) [inline tests/go.bats:66]: nonblocking: how come `GO_VERSION_COMMAND` is necessary?

File: tests/go.bats
Comment (mkenigs) [inline tests/go.bats:75]: nonblocking: it would be good to check a super basic compilation works

File: cli/flox/src/commands/init/go.rs
Comment (mkenigs) [inline cli/flox/src/commands/init/go.rs:431]: blocking: I'm guessing you'll uncomment or drop these tests?

**Review by mkenigs — CHANGES_REQUESTED** [review]
Overall this is looking great! We mostly follow https://conventionalcomments.org/, so basically anything I said was nonblocking you can ignore

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:95]: suggestion(nb): use literal liebreak instead of `\n`
Personally i find mixing `\n` and actual linebreaks harder to imagine printed out.


```suggestion
            * A shell hook to apply environment variables

        ", module_system.get_filename()});
```

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:84]: quib (NB!): I think we should really break this out into a function or out of default impl this in #1266, but no need to worry about this here!

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:143]: suggestion(nb): add a reason to unreachable

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:133]: suggestion(nb): add a reason to unreachable

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:197]:
```suggestion

    /// Detects and returns the possible instance of a Go module or workspace system
```

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:207]:
```suggestion

    /// Returns the provided version obtained from the module system file.
```

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:206]: nothing agains a cows but wouldnt `&'static str` work too? in these impls?

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:289]: todo: refactor Go specific `impl ProvidedVersion` methods into functions or methods that could take and return `ProvidedVersion`.

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs]: I will add an issue as a follow-up implementation once this PR is finished.

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:431]: todo: decide whether to keep or drop commented tests based on coverage results

I commented these tests to check for coverage differences before and after version tests, but the latter weren't working at the time.

File: tests/go.bats
Comment (oxcabe) [reply tests/go.bats:66]: todo: remove leftover `GO_VERSION_COMMAND`

File: tests/go.bats
Comment (oxcabe) [reply tests/go.bats:75]: todo: run `go build` during functional test

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:438]: note: I will ignore this comment until I finish the rest of proposed changes

**Comment by ghudgins** [comment]
ignore my issue, it was a me thing. all things are working for me now on the branch 🚀

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:209]: todo: drop `Cow` wrapper in `get_filename` return type

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:225]: nit (nb): somthing about the naming seems off here `GoModuleSystemMode for GoModuleSystem`
maybe the fact that `GoWorkspaceSystem` and `GoModuleSystem` are both `GoModuleSystemMode`s maybe there are more appropriate names (e.g. GoModSystem/GoWorkSystem?).
Pretty much a non-functional issue so feel free to ingnore.

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:230]: nit(nb): we have a somewhat unwritten convention to take `flox` as the first argument.
Nb, because its non-functional, but it's surprising to see at least

File: cli/flox/src/commands/init/go.rs
Comment (ghudgins) [reply cli/flox/src/commands/init/go.rs:318]: agree here. don't run the hook if we don't (yet) have compatible versions. we're in the process of adding a huge back history of software to Flox so this will hopefully hit a match more often after we make that change

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:225]: note: indeed, took me some time to find a name, I was not expecting it to be the best choice after all

todo: rename `GoModuleSystem -> GoModSystem`, `GoWorkspaceSystem -> GoWorkSystem`

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:230]: note: it makes sense if `flox` is a "core" instance

Comment (ysndr) [reply cli/flox/src/commands/init/go.rs:230]: yeah, `flox` is our global shared context for impure operations basically

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:230]: question: since for the module init hook code, including `InitHook` itself, methods don't have this convention applied, should I leave `Go::new(dir, flox)` as it is?

See:

https://github.com/flox/flox/blob/2a4537aaf4133092a86287373d6f3f04afbbb1c7/cli/flox/src/commands/init/mod.rs#L165-L169

https://github.com/flox/flox/blob/2a4537aaf4133092a86287373d6f3f04afbbb1c7/cli/flox/src/commands/init/mod.rs#L219-L224

Comment (ysndr) [reply cli/flox/src/commands/init/go.rs:230]: Yeah that's the "unwritten" part of that convention ^^

I guess it's better to consistently deviating there although we should probably change it there as well, don't sweat it :)

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:230]: Got it! We'll get there soon

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:318]: note: good catch, I agree with this as well

todo: disable hook running when detected version is incompatible

Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:318]: nb suggestion: users could be warned that the reason for the hook not being run is related to version incompatibility

question: how could I properly phrase such a message? also, could anyone provide me some pointers on how flox manages and emits warnings?

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:324]: question: could we get the default package, check that the version is greater than the `VersionReq` and return it as `Compatible`?

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:384]: todo: split `GoVersion` tests between `from_content` and `parse_content_version_string`

**Comment by oxcabe** [comment]
note: I will currently work on improving the documentation while waiting for review updates

**Comment by ysndr** [comment]
Do you consider this ready for review (still a draft). Or are you just waiting for responses to questions in the threads above before finalizing?

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:246]: suggestion: either the comment is incorrect or it should check if `mod_path` is a directory. currently, if it isnt a file it will return an `Err` i think

```suggestion
    /// This method returns `None` if [GO_MOD_FILENAME] is a directory.
    fn try_new_from_path(flox: &Flox, path: &Path) -> Result<Option<Self>> {
        let mod_path = path.join(GO_MOD_FILENAME);
        if !mod_path.exists() || mod_path.is_dir() {
            return Ok(None);
        }

        let mod_content = fs::read_to_string(mod_path)?;
        Self::try_new_from_content(flox, &mod_content)
    }
```

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [reply cli/flox/src/commands/init/go.rs:325]: Go versions are weird...https://go.dev/doc/toolchain#version

A go version 2 is not to be expected anytime soon, but I agree a comment for posterity might be in order.

@oxcabe I wouldnt worry too much future proofing for go version 2.
1) if there is a major change i think we'll still have the latest go1.* version around that will satisfy a version 1.* constraint.
2) unless there _is_ a version two this "version upgrade" mechanism would be effectively dead code

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:329]: if the `line` starts with whitespace `line.split_whitespace().nth(1)` will be `go` i think (`["", "go", "<version>"][1]`)

```suggestion
            .and_then(|line| line.trim().split_whitespace().nth(1))
```

File: cli/flox/src/commands/init/go.rs
Comment (ysndr) [inline cli/flox/src/commands/init/go.rs:387]: It is at least possible to specify version constraints as `go 1.21rc1` [apparently](https://github.com/search?q=%22go+1.21rc1%22+path%3A**%2Fgo.mod&type=code) which is not valid semver.

That should be commented on as well and maybe have an "expect to fail" test for visibility

File: tests/go/single-dependency/module/go.mod
Comment (ysndr) [inline tests/go/single-dependency/module/go.mod:1]: note (nb): this file is named "single-dependency" but does not declare any dependency 🤔
I guess it "is" the single dependency which makes me wonder if we should have a test that actually imports a 3rd party dependency.

In the python and node(?) tests we import libraries that can be tricky (i.e.e bind to native libs). Not that we have to find something like that for go.

**Review by ysndr — CHANGES_REQUESTED** [review]

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:325]: I'll leave a note as a comment.

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs:391]: Seems to be a rare case, at least.

We could strip anything from the `<version>` string starting from the first alpha char and try to parse and find a compatible version for the remaining string, but it's a bit of a mess IMO.

regex could be the most standard approach to this, but the `regex` crate is not included in the Cargo file and might not be worth adding additional dependencies for such an edge case. Same thing with `nom`, if preferred to `regex`.

File: cli/flox/src/commands/init/go.rs
Comment (oxcabe) [reply cli/flox/src/commands/init/go.rs]: I committed your suggestion and added integration tests covering this check here:
https://github.com/flox/flox/pull/1227/commits/1671e9b7154b4a15f6c815fa61fd362e7b83c536

File: tests/go/single-dependency/module/go.mod
Comment (oxcabe) [reply tests/go/single-dependency/module/go.mod:1]: note: the trickiest one I can think of, based on personal experience, is https://github.com/alexbrainman/odbc/, which requires specific third-party shared libraries.

File: cli/flox/src/commands/init/go.rs
Comment (mkenigs) [reply cli/flox/src/commands/init/go.rs]: You could do `message::warning("whatever")`

File: tests/go.bats
Comment (mkenigs) [inline tests/go.bats:110]: suggestion nonblocking: can these 2 tests be made unit tests? If not they should be moved to `cli/tests`

**Review by mkenigs — APPROVED** [review]
Looks good! I think there are a few nonblocking comments left - do you want to address those, or just merge as is?

**Comment by oxcabe** [comment]
> Looks good! I think there are a few nonblocking comments left - do you want to address those, or just merge as is?

I'll decide some actions for the remaining comments and then we can merge. Thanks for your reviews!

**Review by mkenigs — APPROVED** [review]

**Pull request merged**
